### PR TITLE
Discovery improvements

### PR DESCRIPTION
**Pull request by antonf**

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.discovery.model
cloudferrylib/os/discovery/model.py:309: [W0212(protected-access), 
Model.create] Access to a protected member _original of a client class
cloudferrylib/os/discovery/model.py:312: [W0212(protected-access), 
Model.create] Access to a protected member _original of a client class
cloudferrylib/os/discovery/model.py:405: [W0613(unused-argument), 
Model.load_missing] Unused argument 'cloud'
cloudferrylib/os/discovery/model.py:418: [W0613(unused-argument), 
Model.discover] Unused argument 'cloud'
cloudferrylib/os/discovery/model.py:675: [E1101(no-member), 
Transaction.__enter__] Module 'sqlite3' has no 'connect' member
************* Module fabfile
fabfile.py:269: [W0612(unused-variable), load_yaml_config] Unused variable 
'name'
************* Module tests.cloudferrylib.os.discovery.test_model
tests/cloudferrylib/os/discovery/test_model.py:23: [E1101(no-member), 
UnclosableConnection] Module 'sqlite3' has no 'Connection' member
tests/cloudferrylib/os/discovery/test_model.py:24: [W0613(unused-argument), 
UnclosableConnection.close] Unused argument 'args'
tests/cloudferrylib/os/discovery/test_model.py:148: [E1101(no-member), 
ModelTest.setUp] Module 'sqlite3' has no 'connect' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/2761/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.discovery.model
cloudferrylib/os/discovery/model.py:309: [W0212(protected-access), 
Model.create] Access to a protected member _original of a client class
cloudferrylib/os/discovery/model.py:312: [W0212(protected-access), 
Model.create] Access to a protected member _original of a client class
cloudferrylib/os/discovery/model.py:405: [W0613(unused-argument), 
Model.load_missing] Unused argument 'cloud'
cloudferrylib/os/discovery/model.py:418: [W0613(unused-argument), 
Model.discover] Unused argument 'cloud'
cloudferrylib/os/discovery/model.py:675: [E1101(no-member), 
Transaction.__enter__] Module 'sqlite3' has no 'connect' member
************* Module tests.cloudferrylib.os.discovery.test_model
tests/cloudferrylib/os/discovery/test_model.py:23: [E1101(no-member), 
UnclosableConnection] Module 'sqlite3' has no 'Connection' member
tests/cloudferrylib/os/discovery/test_model.py:24: [W0613(unused-argument), 
UnclosableConnection.close] Unused argument 'args'
tests/cloudferrylib/os/discovery/test_model.py:148: [E1101(no-member), 
ModelTest.setUp] Module 'sqlite3' has no 'connect' member
************* Module fabfile
fabfile.py:269: [W0612(unused-variable), load_yaml_config] Unused variable 
'name'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.discovery.model
cloudferrylib/os/discovery/model.py:309: [W0212(protected-access), 
Model.create] Access to a protected member _original of a client class
cloudferrylib/os/discovery/model.py:312: [W0212(protected-access), 
Model.create] Access to a protected member _original of a client class
cloudferrylib/os/discovery/model.py:405: [W0613(unused-argument), 
Model.load_missing] Unused argument 'cloud'
cloudferrylib/os/discovery/model.py:418: [W0613(unused-argument), 
Model.discover] Unused argument 'cloud'
cloudferrylib/os/discovery/model.py:675: [E1101(no-member), 
Transaction.__enter__] Module 'sqlite3' has no 'connect' member
************* Module tests.cloudferrylib.os.discovery.test_model
tests/cloudferrylib/os/discovery/test_model.py:23: [E1101(no-member), 
UnclosableConnection] Module 'sqlite3' has no 'Connection' member
tests/cloudferrylib/os/discovery/test_model.py:24: [W0613(unused-argument), 
UnclosableConnection.close] Unused argument 'args'
tests/cloudferrylib/os/discovery/test_model.py:148: [E1101(no-member), 
ModelTest.setUp] Module 'sqlite3' has no 'connect' member
************* Module fabfile
fabfile.py:269: [W0612(unused-variable), load_yaml_config] Unused variable 
'name'

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/os/discovery/model.py [hunk 1]
```diff
@@ -419,29 +518,45 @@ class Reference(fields.Field):
...
             return None
         model = self.model_class
         pk_field = model.pk_field
         if self.many:
-            return [_object_id_to_dict(model, getattr(x, pk_field))
+            return [getattr(x, pk_field).to_dict(model)
                     for x in value]
         else:
-            return _object_id_to_dict(model, getattr(value, pk_field))
+            return getattr(value, pk_field).to_dict(model)
```
Comment (roman-verchikov) [inline cloudferrylib/os/discovery/model.py]: Method which returns different types of objects (list of objects, and single object) is somewhat misleading, and feels like a "code smell" to me, please make sure returning these is expected. Also I was unable to find any usages of `_serialize()` and `_deserialize()`, are these really needed?

Comment (antonf) [inline cloudferrylib/os/discovery/model.py]: It is custom Marshmallow field, `_serialize` and `_deserialize` are designed to work with JSON-like objects (lists and dictionaries fall into this category). More info: http://marshmallow.readthedocs.org/en/latest/custom_fields.html

**Comment by roman-verchikov** [comment]
+1, can be merged

**Pull request merged**